test/ops/test_provenance.py

```diff
@@ -1,21 +1,86 @@
 # Copyright Contributors to the Pyro project.
 # SPDX-License-Identifier: Apache-2.0
 
-import numpy as np
+import inspect
+
 import pytest
 
-from numpyro.ops.provenance import ProvenanceArray, eval_provenance, get_provenance
+import jax
+from jax.api_util import flatten_fun_nokwargs
+import jax.core as core
+import jax.linear_util as lu
+import jax.numpy as jnp
+
+from numpyro.ops.provenance import eval_provenance
 
 
 @pytest.mark.parametrize(
     "f, inputs, expected_output",
     [
-        (lambda x, y: x + y, ({"a"}, {}), {"a"}),
-        (lambda x, y: x + y, ({"a"}, {"b"}), {"a", "b"}),
-        (lambda x, y: x + y, ({"a", "c"}, {"a", "b"}), {"a", "b", "c"}),
-        (lambda x, y, z: x + y, ({"a"}, {"b"}, {"c"}), {"a", "b"}),
+        (lambda a, b: a + 1, ("a", "b"), {"a"}),
+        (lambda a, b: jax.scipy.special.xlogy(a, b), ("a", "b"), {"a", "b"}),
+        (lambda a, b, c: a + b, ("a", "b", "c"), {"a", "b"}),
+        (
+            lambda a, b: {"sum": a + b, "zero": 0},
+            ("a", "b"),
+            {"sum": {"a", "b"}, "zero": set()},
+        ),
     ],
 )
 def test_provenance(f, inputs, expected_output):
-    inputs = [ProvenanceArray(np.array([0]), p) for p in inputs]
-    assert get_provenance(eval_provenance(f, *inputs)) == expected_output
+    inputs = {p: 0 for p in inspect.getfullargspec(f).args}
+    assert eval_provenance(f, **inputs) == expected_output
+
+
+def test_provenance_const():
+    def f(x):
+        with jax.ensure_compile_time_eval():
+            y = jnp.ones(4)
+        return x + y
+
+    jaxpr = jax.make_jaxpr(f)(jnp.zeros((3, 4), jnp.float32))
+    assert len(jaxpr.consts) == 1
+    assert eval_provenance(f, x=3) == {"x"}
+
+
+def test_provenance_fori():
+    def f(x, y, z):
+        del z
```

Comment:
Why do you have to delete `z` here?

Reply:
we don't use z in this function. it is just a good practice i think.